velox/exec/MergeSource.cpp

```diff
@@ -91,20 +170,17 @@ class LocalMergeSource : public MergeSource {
     }
 
    private:
-    void notifyConsumers() {
-      for (auto& promise : consumerPromises_) {
-        promise.setValue();
-      }
-      consumerPromises_.clear();
+    void notifyConsumers(PromiseNotifier& notifier) {
+      notifier.add(std::move(consumerPromises_));
+      VELOX_CHECK(consumerPromises_.empty());
     }
 
-    void notifyProducers() {
-      for (auto& promise : producerPromises_) {
-        promise.setValue();
-      }
-      producerPromises_.clear();
+    void notifyProducers(PromiseNotifier& notifier) {
+      notifier.add(std::move(producerPromises_));
+      VELOX_CHECK(producerPromises_.empty());
```

Comment:
Why do we want to set promises outside instead of in-place?

Reply:
To avoid the continuation inside the lock